x/dsmr/storage.go

```diff
@@ -168,7 +178,18 @@ func (s *ChunkStorage[T]) init() error {
 			return fmt.Errorf("failed to parse chunk %s at slot %d", chunkID, slot)
 		}
 		s.chunkEMap.Add([]emapChunk[T]{{chunk: chunk}})
-		s.chunkMap[chunk.id] = &StoredChunkSignature[T]{Chunk: chunk}
+		storedChunkSig := &StoredChunkSignature[T]{Chunk: chunk}
+		s.pendingChunkMap[chunk.id] = storedChunkSig
+		s.pendingChunksProducers[chunk.Producer] = append(s.pendingChunksProducers[chunk.Producer], storedChunkSig)
+	}
+
+	// sort all the producer's chunks.
+	// no need to write back to the pendingChunksProducers since all the slice
+	// changes are in-place.
+	for _, chunks := range s.pendingChunksProducers {
+		sort.Slice(chunks, func(i, j int) bool {
+			return chunks[i].Chunk.Expiry < chunks[j].Chunk.Expiry
+		})
```

Comment:
Would it make sense to use the `ExpiryHeap` data structure here?

Reply:
I think that using a heap here isn't really what we're after, as we want to be able to dynamically "peek" into random windows. What we could do, is to store a slice of `StoredChunkWindow` as
```golang
type StoredChunkWindow struct {
    *StoredChunkSignature[T]
    windowWeight uint64
}
```
this would reduce the O(n^2) complexity in `CheckRateLimit` into `O(n)`.

Unfortunately, this is more theoretical than practical. Given a reasonable time limits on the time validity window, the number of chunks that would be in question above would be quite small, making this optimization less meaningful.

Reply:
As discussed offline, it should be alright to allow the window to be the same as the validity window, which will allow us to only track a sum of total bandwidth consumed within that window and not worry about guaranteeing there's never a shorter interval where a validator consumes more bandwidth than they are allotted for that period of time.